fsmonitor updates for improved performance

kewillford opened the pull request:
This change does two main things.
1. Adds a check for the `CE_FSMONITOR_VALID` flag in the `ce_uptodate` macro so that whenever the code is checking if any entry is up to date the fsmonitor flag will be taken into consideration.
2. When unpacking trees keep the fsmonitor data so that the next command does not have to pay the price to check all the entries.

kewillford commented:
Here are some of the performance differences with these changes

command|Previous Duration|Current Duration|Seconds difference|percent change
-|-|-|-|-|
checkout after folderplaceholder enumeration|17.40225|3.21683|-14.18542|-81.51486
status 1 after checkout feature/gvfs/perftest/defaultBranch|15.53006|3.03953|-12.49053|-80.42809
status 1 after checkout -b user/me/topic feature/gvfs/perftest/defaultBranch~1|15.66574|3.09206|-12.57368|-80.26228
status 1 after pull --ff-only origin feature/gvfs/perftest/checkoutBranch|16.08161|3.62207|-12.45954|-77.47694
status 1 after checkout -b pullTest feature/gvfs/perftest/checkoutBranch~10|15.61599|3.63645|-11.97954|-76.71329
status 1 after merge feature/gvfs/perftest/defaultBranch --no-commit|13.1538|3.17194|-9.98186|-75.88575
git reset --hard HEAD|26.76172|6.68994|-20.07178|-75.00183
status 1 after reset --hard HEAD|12.11019|3.15032|-8.95987|-73.98621
status 1 after checkout feature/gvfs/perftest/checkoutBranch|17.4508|4.88049|-12.57031|-72.03286
status 1 after stash|10.89549|3.09768|-7.79781|-71.56915
git add --all|30.97013|10.36075|-20.60938|-66.54599
status 2 after First ReadFiles|4.35711|1.60531|-2.7518|-63.15654
git stash|31.56066|11.86821|-19.69245|-62.39556
git merge feature/gvfs/perftest/defaultBranch --no-commit|21.15363|8.60551|-12.54812|-59.31899
git stash pop|19.23136|9.56369|-9.66767|-50.27034
git rebase feature/gvfs/perftest/defaultBranch|26.31549|13.19811|-13.11738|-49.84661
status 1 after merge --abort|14.68546|7.55865|-7.12681|-48.5297

derrickstolee commented:
We need to figure something out about how fsmonitor talks specifically to watchman. We are not robust to script-level frequency (my test is on v2.24.0-rc2).

The `GIT_TEST_FSMONITOR` environment variable can take a hook path, and there is an included hook for Watchman. There are numerous issues with this integration on Linux (we cannot delete repos after registering them with Watchman, so that causes many test failures), but also even the simple `test_commit` function doesn't work!

For example:

```
GIT_TRACE=1 GIT_TEST_FSMONITOR="$(pwd)/t7519/fsmonitor-watchman" ./t7060-wtstatus.sh -x -v -d -i
```

In test 5, the following commands are run in order:

```
		test_commit initial foo "" &&
		test_commit modify foo foo &&
```

and here are the logs for those two lines:

```
+ test_commit initial foo 
+ notick=
+ signoff=
+ indir=
+ test 3 != 0
+ break
+ indir=
+ file=foo
+ echo 
+ git add foo
trace: built-in: git add foo
trace: run_command: cd '/_git/git/t/trash directory.t7060-wtstatus/mdconflict'; /_git/git/t/t7519/fsmonitor-watchman 1 1572618499310242307
Adding '/_git/git/t/trash directory.t7060-wtstatus/mdconflict' to watchman's watch list.
+ test -z 
+ test_tick
+ test -z set
+ test_tick=1112912173
+ GIT_COMMITTER_DATE=1112912173 -0700
+ GIT_AUTHOR_DATE=1112912173 -0700
+ export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
+ git commit -m initial
trace: built-in: git commit -m initial
trace: run_command: cd '/_git/git/t/trash directory.t7060-wtstatus/mdconflict'; /_git/git/t/t7519/fsmonitor-watchman 1 1572618499374543700
trace: run_command: git gc --auto
trace: built-in: git gc --auto
[master (root-commit) a3c5375] initial
 Author: A U Thor <author@example.com>
 1 file changed, 1 insertion(+)
 create mode 100644 foo
+ git tag initial
trace: built-in: git tag initial
+ test_commit modify foo foo
+ notick=
+ signoff=
+ indir=
+ test 3 != 0
+ break
+ indir=
+ file=foo
+ echo foo
+ git add foo
trace: built-in: git add foo
trace: run_command: cd '/_git/git/t/trash directory.t7060-wtstatus/mdconflict'; /_git/git/t/t7519/fsmonitor-watchman 1 1572618499425203804
+ test -z 
+ test_tick
+ test -z set
+ test_tick=1112912233
+ GIT_COMMITTER_DATE=1112912233 -0700
+ GIT_AUTHOR_DATE=1112912233 -0700
+ export GIT_COMMITTER_DATE GIT_AUTHOR_DATE
+ git commit -m modify
trace: built-in: git commit -m modify
trace: run_command: cd '/_git/git/t/trash directory.t7060-wtstatus/mdconflict'; /_git/git/t/t7519/fsmonitor-watchman 1 1572618499517349606
On branch master
nothing to commit, working tree clean
```

The second `git commit -m modify` fails because the second `git add foo` did nothing. It triggered the watchman call, but did not receive a change for that file, so did not update the index. (I verified this by adding an extra `git status` call in the `test_commit` code and re-running the test.)

We will need to think about how to make our Watchman integration more robust and set up some automation to run the test suite with Watchman specifically.

cc: @kewillford, @jrbriggs, @wilbaker, @dscho

jeffhostetler commented:
@derrickstolee    Can you run this with GIT_TRACE_FSMONITOR turned on too?  Not sure if that'll help or not, but worth a shot.

D'oh, I just noticed that you did have that turned on in the above command line.  It there anything in that other log file?

dscho commented on t/t7113-post-index-change-hook.sh:
```diff
@@ -80,9 +80,6 @@ test_expect_success 'test status, add, commit, others trigger hook without flags
 	test_path_is_missing testfailure &&
-	git update-index &&
-	test_path_is_missing testsuccess &&
-	test_path_is_missing testfailure &&
```
Since this test does not use the fsmonitor, it is a bit confusing to a reviewer such as myself why this test needs to be adjusted.

Could you please add an explanation to the commit message, specifically for this test's update?

Oh, and please add your [sign-off](https://git-scm.com/docs/SubmittingPatches#signoff)...

dscho commented on t/t7519-status-fsmonitor.sh:
```diff
@@ -106,6 +106,8 @@ EOF
 # test that "update-index --fsmonitor-valid" sets the fsmonitor valid bit
 test_expect_success 'update-index --fsmonitor-valid" sets the fsmonitor valid bit' '
+	write_script .git/hooks/fsmonitor-test<<-\EOF &&
+	EOF
```
Isn't the integration script (i.e. `fsmonitor-test`) initialized only two test cases further down?

I guess this change is necessary because the fsmonitor was not exercised in this test case without this current patch. If that is the case, I would prefer to initialize `fsmonitor_test` as an empty script in the `setup` test case already.

dscho commented on t/t7519-status-fsmonitor.sh:
```diff
@@ -164,6 +166,8 @@ EOF
 # test that newly added files are marked valid
 test_expect_success 'newly added files are marked valid' '
+	write_script .git/hooks/fsmonitor-test<<-\EOF &&
+	EOF
```
I guess I do not understand this diff hunk. previously, we let the test fsmonitor tell Git to look at the new files, but now we don't?

Is this a diff hunk that is not exactly necessary for the test case to pass, but merely to make the test more accurate, by _disallowing_ fsmonitor to trigger `lstat()`s on the three new files and instead forcing it to rely on the implicit information given by the git add` commands?

kewillford replied:
Thanks for pointing this out.  The issue is that with these changes the fsmonitor data is refreshed when any command reads the index. Left unchanged the `git ls-files` would refresh and be using the dirty files from the previous test which would include the newly added files and they would be marked as dirty which would fail this test. Perhaps I can use the `test-tool dump-fsmonitor` to get the bitmap to compare against instead of using `ls-files`.

wilbaker replied:
Follow up question here: what behavior is this test specifically trying to validate?  

The comment says:

`test that newly added files are marked valid`

But based on the change you made it seems like newly added files will not always be marked as valid (i.e. if they were dirty before they will still be considered dirty after the `git add`).

Is there there something I'm missing (e.g. was the test `.git/hooks/fsmonitor-test` behaving incorrectly, and that's why it needs to be set to empty at the start of the test)?

wilbaker replied:
Follow up question here: what behavior is this test specifically trying to validate?  

The comment says:

`test that newly added files are marked valid`

But based on the change you made it seems like newly added files will not always be marked as valid (i.e. if they were dirty before they will still be considered dirty after the `git add`).

Is there there something I'm missing (e.g. was the test `.git/hooks/fsmonitor-test` behaving incorrectly, and that's why it needs to be set to empty at the start of the test)?

wilbaker replied:
Follow up question here: what behavior is this test specifically trying to validate?  

The comment says:

`test that newly added files are marked valid`

But based on the change you made it seems like newly added files will not always be marked as valid (i.e. if they were dirty before they will still be considered dirty after the `git add`).

Is there there something I'm missing (e.g. was the test `.git/hooks/fsmonitor-test` behaving incorrectly, and that's why it needs to be set to empty at the start of the test)?

kewillford replied:
Maybe for this test the 
```
write_script .git/hooks/fsmonitor-test<<-\EOF &&
EOF
```

needs to be before the `git ls-files` so that the `git add` runs with the paths in `.git/hooks/fsmonitor-test` but for each `git add` the single path for the added file would need to be in `.git/hooks/fsmonitor-test` otherwise the last `git add` would mark the other paths as dirty and save that index out.

I'm taking a closer look at the tests because before the content of the `.git/hooks/fsmonitor-test` did not affect commands that were not refreshing cache entries whereas now any commands that reads the index will get the entries marked dirty that are in the `.git/hooks/fsmonitor-test`.  This means that even `git ls-files` which is being used to validate the CE_FSMONITOR_VALID could have that affected by what is in `.git/hooks/fsmonitor-test`.  So if left with all the paths in `.git/hooks/fsmonitor-test`, those paths will be dirty for every git command that reads the index.

So in most cases we need to make sure `.git/hooks/fsmonitor-test` is empty for any validating commands like `git ls-files`  so that the contents of that file will not change what cache entries are dirty.

I might also try using the `test-dump-fsmonitor` since that it only dumping the index entries before applying the bitmap or the `.git/hooks/fsmonitor-test` paths.

wilbaker replied:
> or each git add the single path for the added file would need to be in .git/hooks/fsmonitor-test otherwise the last git add would mark the other paths as dirty and save that index out.

Ahh, this is the part I was missing, thanks!

dscho commented on t/t7519-status-fsmonitor.sh:
```diff
@@ -218,11 +222,12 @@ test_expect_success '*only* files returned by the integration script get flagged
 	dirty_repo &&
 	git add . &&
+	write_script .git/hooks/fsmonitor-test<<-\EOF &&
+	EOF
```
So basically the change in `'newly added files are marked valid'` made it so that the fsmonitor is not allowed to tell Git to look at any file's stat in all the test cases up until here. Hmm. I would really like to have at least a paragraph in the commit message providing a compelling argument why that is a good thing.

And then I really don't understand why we have to have the full `fsmonitor-test` script again, but only for `dirty_repo` and for `git add .`?

dscho commented on unpack-trees.c:
```diff
@@ -1581,6 +1581,9 @@ int unpack_trees(unsigned len, struct tree_desc *t, struct unpack_trees_options
 
+	if (o->src_index->fsmonitor_last_update)
+		o->result.fsmonitor_last_update = o->src_index->fsmonitor_last_update;
+
```
This probably makes sense on its own, in a separate patch (which I would move to the front of the PR, if I were you). It should not break any tests, or does it?

dscho commented on fsmonitor.c:
```diff
@@ -55,10 +55,11 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	istate->fsmonitor_dirty = fsmonitor_dirty;
 
-	if (istate->fsmonitor_dirty->bit_size > istate->cache_nr)
+	if (!istate->split_index && istate->fsmonitor_dirty->bit_size > istate->cache_nr)
```
Makes sense.

dscho commented on fsmonitor.c:
```diff
@@ -55,10 +55,11 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 		BUG("fsmonitor_dirty has more entries than the index (%"PRIuMAX" > %"PRIuMAX")",
 		    (uintmax_t)istate->fsmonitor_dirty->bit_size, (uintmax_t)istate->cache_nr);
 
+
```
Let's not introduce that empty line here, unless you really care about it?

dscho reviewed (approved):
Looks good!

I asked for a few clarifications, and suggest to split the change that copies the `last_update` when copying an index into its own commit, but none of these are super-important.

But please add your [sign-off](https://git-scm.com/docs/SubmittingPatches#signoff) to the commit messages.

wilbaker commented on fsmonitor.c:
```diff
@@ -83,7 +84,7 @@ void write_fsmonitor_extension(struct strbuf *sb, struct index_state *istate)
 	int fixup = 0;
 
-	if (istate->fsmonitor_dirty->bit_size > istate->cache_nr)
+	if (!istate->split_index && istate->fsmonitor_dirty->bit_size > istate->cache_nr)
```
1. Is a similar change needed in `tweak_fsmonitor`?  If so, I think that would point to a bigger issue as `ewah_each_bit(istate->fsmonitor_dirty, fsmonitor_ewah_callback, istate);` can `BUG` if called on a split_index.

2. This `BUG` (and the one in `read_fsmonitor_extension`) were introduced in git 2.24 in commit https://github.com/git/git/commit/3444ec2eb2be58c285d2bf04f39e6e9ea5eda9a2.  Should the commit with this fix be fast-tracked to the mailing list as its own patch?

cc: @dscho for his thoughts

dscho replied:
Hmm. I'll have to free up some brain cycles to think about this. So close to v2.24.0, I don't think we will get this into `master`, though. It only affects the split index feature, though, right? That's a fringe feature (and not a particularly well-designed one, I think), so I don't think that we will have to rush things with those fixes.

wilbaker replied:
> It only affects the split index feature, though, right?

Yep this issue is specific to split-index + fsmonitor, thanks!

wilbaker reviewed (approved):
Changes look good, same questions as @dscho on the updates to the tests.

derrickstolee commented:
@kewillford when you get this rebased on top of `features/sparse-checkout-2.24.0` then I can launch the C# tests using watchman for additional confidence.

derrickstolee commented:
@kewillford I kicked off a new build for you [here](https://dev.azure.com/gvfs/ci/_build/results?buildId=17538&view=results).

kewillford committed:
fsmonitor: add script for debugging and update script for tests

The fsmonitor script that can be used for running all the git tests
using watchman was causing some of the tests to fail because it wrote
to stderr and created some files for debugging purposes.

Add a new debug script to use with debugging and modify the other script
to remove the code that would cause tests to fail.

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

kewillford committed:
fsmonitor: do not compare bitmap size with size of split index

3444ec2e ("fsmonitor: don't fill bitmap with entries to be removed",
2019-10-11) added a handful of sanity checks that make sure that a
bit position in fsmonitor bitmap does not go beyond the end of the
index.  As each bit in the bitmap corresponds to a path in the
index, this is the right check most of the time.

Except for the case when we are in the split-index mode and looking
at a delta index that is to be overlayed on the base index but
before the base index has actually been merged in, namely in read_
and write_fsmonitor_extension().  In these codepaths, the entries in
the split/delta index is typically a small subset of the entire set
of paths (otherwise why would we be using split-index?), so the
bitmap used by the fsmonitor is almost always larger than the number
of entries in the partial index, and the incorrect comparison would
trigger the BUG().

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

kewillford committed:
fsmonitor: check CE_FSMONITOR_VALID in ce_uptodate

When using fsmonitor the CE_FSMONITOR_VALID flag should be checked when
wanting to know if the entry has been updated. If the flag is set the
entry should be considered up to date and the same as if the CE_UPTODATE
is set.

In order to trust the CE_FSMONITOR_VALID flag, the fsmonitor data needs to
be refreshed when the fsmonitor bitmap is applied to the index in
tweak_fsmonitor. Since the fsmonitor data is kept up to date for every
command, some tests needed to be updated to take that into account.

istate->untracked->use_fsmonitor was set in tweak_fsmonitor when the
fsmonitor bitmap data was loaded and is now in refresh_fsmonitor since
that is being called in tweak_fsmonitor. refresh_fsmonitor will only be
called once and any other callers should be setting it when refreshing
the fsmonitor data so that code can use the fsmonitor data when checking
untracked files.

When writing the index, fsmonitor_last_update is used to determine if
the fsmonitor bitmap should be created and the extension data written to
the index. When running through unpack-trees this is not copied to the
result index. This makes the next time a git command is ran do all the
work of lstating all files to determine what is clean since all entries
in the index are marked as dirty since there wasn't any fsmonitor data
saved in the index extension.

Copying the fsmonitor_last_update to the result index will cause the
extension data for fsmonitor to be in the index for the next git command
to use.

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>
